File: construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        preorderIdx = [0]
        def buildTreeHelper(preorderIdx,left,right):
            if left>right:
                return None
            
            node = TreeNode(preorder[preorderIdx[0]])
            idx = search(preorder[preorderIdx[0]],left,right)
            preorderIdx[0] += 1
            if left == right:
                return node
            
            node.left = buildTreeHelper(preorderIdx,left,idx-1)
            node.right = buildTreeHelper(preorderIdx,idx+1,right)
            return node
        
        def search(target,left,right):
            for i in range(left,right+1):
                if inorder[i] == target:
                    return i
            return -1
        
        return buildTreeHelper(preorderIdx,0,len(preorder)-1)
            

File: construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public static TreeNode buildTreeHelper(int[] preorder, int[] inorder, int[] idx,int left, int right){
        if(left>right){
            return null;
        }
        
        TreeNode node = new TreeNode(preorder[idx[0]]);
        int inorderIdx = searchIdx(inorder,left,right,preorder[idx[0]]);
        idx[0]++;
        if(left == right){
            return node;
        }
        
        node.left = buildTreeHelper(preorder,inorder,idx,left,inorderIdx-1);
        node.right = buildTreeHelper(preorder,inorder,idx,inorderIdx+1,right);
        return node;
        
    }
    
    public static int searchIdx(int[] inorder, int left, int right, int target){
        
        for(int i=left;i<=right;i++){
            if(inorder[i] == target){
                return i;
            }
        }
        return -1;
    }
    
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int[] idx = new int[1];
        idx[0] = 0;
        return buildTreeHelper(preorder,inorder,idx,0,preorder.length-1);
    }
}